warehouse/internal/repo/staging.go

```diff
@@ -0,0 +1,288 @@
+	SourceTaskRunID       string `json:"source_task_run_id"`
+	SourceJobID           string `json:"source_job_id"`
+	SourceJobRunID        string `json:"source_job_run_id"`
+	TimeWindowYear        int    `json:"time_window_year"`
+	TimeWindowMonth       int    `json:"time_window_month"`
+	TimeWindowDay         int    `json:"time_window_day"`
+	TimeWindowHour        int    `json:"time_window_hour"`
+	DestinationRevisionID string `json:"destination_revision_id"`
+}
+
+func metadataFromStagingFile(stagingFile *model.StagingFile) metadataSchema {
+	return metadataSchema{
+		UseRudderStorage:      stagingFile.UseRudderStorage,
+		SourceBatchID:         stagingFile.SourceBatchID,
+		SourceTaskID:          stagingFile.SourceTaskID,
+		SourceTaskRunID:       stagingFile.SourceTaskRunID,
+		SourceJobID:           stagingFile.SourceJobID,
+		SourceJobRunID:        stagingFile.SourceJobRunID,
+		TimeWindowYear:        stagingFile.TimeWindow.Year(),
+		TimeWindowMonth:       int(stagingFile.TimeWindow.Month()),
+		TimeWindowDay:         stagingFile.TimeWindow.Day(),
+		TimeWindowHour:        stagingFile.TimeWindow.Hour(),
+		DestinationRevisionID: stagingFile.DestinationRevisionID,
+	}
+}
+
+func (m *metadataSchema) SetStagingFile(stagingFile *model.StagingFile) {
+	stagingFile.UseRudderStorage = m.UseRudderStorage
+	stagingFile.SourceBatchID = m.SourceBatchID
+	stagingFile.SourceTaskID = m.SourceTaskID
+	stagingFile.SourceTaskRunID = m.SourceTaskRunID
+	stagingFile.SourceJobID = m.SourceJobID
+	stagingFile.SourceJobRunID = m.SourceJobRunID
+	stagingFile.TimeWindow = time.Date(m.TimeWindowYear, time.Month(m.TimeWindowMonth), m.TimeWindowDay, m.TimeWindowHour, 0, 0, 0, time.UTC)
+	stagingFile.DestinationRevisionID = m.DestinationRevisionID
+}
+
+func (repo *StagingFiles) init() {
+	repo.once.Do(func() {
+		if repo.Now == nil {
+			repo.Now = timeutil.Now
+		}
+	})
+}
+
+// Insert inserts a staging file into the staging files table. It returns the ID of the inserted staging file.
+//
+// NOTE: The following fields are ignored and set by the database:
+// - ID
+// - Error
+// - CreatedAt
+// - UpdatedAt
+func (repo *StagingFiles) Insert(ctx context.Context, stagingFile *model.StagingFile) (int64, error) {
+	repo.init()
+
+	var (
+		id                        int64
+		firstEventAt, lastEventAt interface{}
+	)
+
+	firstEventAt = stagingFile.FirstEventAt.UTC()
+	if stagingFile.FirstEventAt.IsZero() {
+		firstEventAt = nil
+	}
+
+	lastEventAt = stagingFile.LastEventAt.UTC()
+	if stagingFile.LastEventAt.IsZero() {
+		lastEventAt = nil
+	}
+
+	m := metadataFromStagingFile(stagingFile)
+	rawMetadata, err := json.Marshal(&m)
+	if err != nil {
+		return id, fmt.Errorf("marshaling metadata: %w", err)
+	}
+	now := repo.Now()
+
+	schemaPayload, err := json.Marshal(stagingFile.Schema)
+	if err != nil {
+		return id, fmt.Errorf("marshaling schema: %w", err)
+	}
+
+	err = repo.DB.QueryRowContext(ctx,
+		`INSERT INTO `+tableName+` (
```

Comment:
should we move tablename at repo itself due to close association between the two ?

Reply:
Do you mean adding it as part of the struct instead of a const at the repo level?